src/Spectre.libDataset/IReadOnlyDataset.h

```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include <gsl.h>
+
+namespace Spectre
+{
+    template<bool, typename T>
+    struct deduce_helper
+    {
+        using type = typename T;
+    };
+
+    template<typename T>
+    struct deduce_helper<false, T>
+    {
+        using type = typename const T&;
+    };
+
+    template<typename T>
+    struct deduce : deduce_helper<std::is_fundamental<T>::value || std::is_enum<T>::value || std::is_pointer<T>::value, T>
```

Comment:
Seems like this one is destructing [this line](https://github.com/spectre-team/spectre/pull/94/files#diff-94d41ff8ccd535c63a6f4b84f58ebdc3R34). Maybe we should drop it - const ref to e.g. `int` is of the same weight as copy.